src/util/include/util/SmallVector.h

```diff
@@ -0,0 +1,115 @@
+        } else {
+            dataVector = {list.begin(), list.end()};
+        }
+    }
+
+    size_type size() const { return nb; }
+    bool empty() const { return nb == 0; }
+
+    T& operator[](size_type pos) { return nb <= N ? dataArray[pos] : dataVector[pos]; }
+    const T& operator[](size_type pos) const { return nb <= N ? dataArray[pos] : dataVector[pos]; }
+    T& front() { return nb <= N ? dataArray.front() : dataVector.front(); }
+    const T& front() const { return nb <= N ? dataArray.front() : dataVector.front(); }
+    T& back() { return nb <= N ? dataArray[nb - 1] : dataVector.back(); }
+    const T& back() const { return nb <= N ? dataArray[nb - 1] : dataVector.back(); }
+
+    T* data() { return nb <= N ? dataArray.data() : dataVector.data(); }
+    const T* data() const { return nb <= N ? dataArray.data() : dataVector.data(); }
+
+    using const_iterator = const T*;
+    using iterator = T*;
+
+    iterator begin() { return nb <= N ? dataArray.data() : dataVector.data(); }
+    const_iterator begin() const { return nb <= N ? dataArray.data() : dataVector.data(); }
+    iterator end() { return nb <= N ? dataArray.data() + nb : dataVector.data() + nb; }
+    const_iterator end() const { return nb <= N ? dataArray.data() + nb : dataVector.data() + nb; }
+
+    void push_back(const T& value) {
+        if (nb < N) {
+            dataArray[nb] = value;
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.push_back(value);
+        } else {
+            dataVector.push_back(value);
+        }
+        ++nb;
+    }
+    void push_back(T&& value) {
+        if (nb < N) {
+            dataArray[nb] = std::move(value);
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.push_back(std::move(value));
+        } else {
+            dataVector.push_back(std::move(value));
+        }
+        ++nb;
+    }
+    template <class... Args>
+    void emplace_back(Args&&... args) {
+        if (nb < N) {
+            dataArray[nb] = T(std::forward<Args>(args)...);
+        } else if (nb == N) {
+            dataVector.reserve(N + 1);
+            std::move(dataArray.begin(), dataArray.end(), std::back_inserter(dataVector));
+            dataVector.emplace_back(std::forward<Args>(args)...);
+        } else {
+            dataVector.emplace_back(std::forward<Args>(args)...);
+        }
+        ++nb;
+    }
+    void pop_back() {
+        if (nb > N) {
+            if (nb == N + 1) {
+                std::move(dataVector.begin(), std::prev(dataVector.end()), dataArray.begin());
+                dataVector.clear();
+            } else {
+                dataVector.pop_back();
+            }
+        }
+        --nb;
+    }
```

Comment:
**Minor nit:** I know this is way easier, but keeping the data in the vector might be faster (hardly depends on N).
E.g. when having N*sizeof(T) is less than one cache line, this makes probably no difference.
Also, dataVector.clear(); is faster, because it does nothing (besides a change of the end-iterator). If you planned to free memory, you should re-assign an empty vector: `dataVector = std::vector<value_type>(0);`.

I'll leave that to you.

Reply:
I find it easier to keep the invariant "if nb <= N then the data is in the stack". Also less confusing for the future user of this class. I'll keep this behaviour (with the suggested `dataVector = std::vector<value_type>(0);`).